crates/swc/src/plugin.rs

```diff
@@ -42,27 +42,38 @@ struct RustPlugins {
 
 impl RustPlugins {
     #[cfg(feature = "plugin")]
-    fn apply(&mut self, mut n: Program) -> Result<Program, anyhow::Error> {
+    fn apply(&mut self, n: Program) -> Result<Program, anyhow::Error> {
         use anyhow::Context;
+        use swc_common::plugin::{deserialize_from_bytes, serialize_into_bytes};
 
+        let mut serialized = serialize_into_bytes(&n)?;
+
+        // Run plugin transformation against current program.
+        // We do not serialize / deserialize between each plugin execution but
+        // copies raw transformed bytes directly into plugin's memory space.
+        // Note: This doesn't mean plugin won't perform any se/deserialization: it
+        // still have to construct from raw bytes internally to perform actual
+        // transform.
         if let Some(plugins) = &self.plugins {
             for p in plugins {
                 let config_json = serde_json::to_string(&p.1)
                     .context("failed to serialize plugin config as json")?;
 
                 let path = swc_plugin_runner::resolve::resolve(&p.0)?;
 
-                n = swc_plugin_runner::apply_js_plugin(
+                serialized = swc_plugin_runner::apply_js_plugin(
```

Comment:
This is the part host skips serialize / deserialize: We do not need to pass `Program` into each plugin since each plugin invocation need to serialize & copy over into guest anyway: instead we supply previous (or initial) raw bytes which can be copied into guest directly.